File: backend/src/controllers/FavoriteController.js
const Favorite = require('../models/Favorite')
const Project = require('../models/Project')
const Student = require('../models/Student')
const Employer = require('../models/Employer')

class FavoriteController {
  // [POST] /favorites
  async saveFavorite(req, res, next) {
    try {
      const accountId = req.account._id
      const { projectId } = req.body

      const project = await Project.findById(projectId)
      if (!project) {
        return res.status(404).json({ message: 'Project not found' })
      }

      const existingFavorite = await Favorite.findOne({ account: accountId, project: projectId })
      if (existingFavorite) {
        return res.status(400).json({ message: 'This project is already in your favorites' })
      }

      const favorite = await Favorite.create({ account: accountId, project: projectId })
      res.status(201).json({ message: 'Project saved to favorites successfully', favorite })
    } catch (err) {
      res.status(500).json({ message: 'Error saving favorite', error: err.message })
    }
  }

  // [GET] /favorites
  async getFavorites(req, res, next) {
    try {
      const accountId = req.account._id

      // Lấy danh sách project còn tồn tại
      const existingProjects = await Project.find({}).select('_id')
      const existingProjectIds = new Set(existingProjects.map(p => p._id.toString()))

      const favorites = await Favorite.find({ account: accountId })
        .populate({
          path: 'project',
          select: '-__v',
          populate: { path: 'account', select: 'role avatar' }
        })
        .sort({ createdAt: -1 })

      // Lọc ra các favorite có project tồn tại trong project data
      const validFavorites = favorites.filter(
        fav => fav.project && existingProjectIds.has(fav.project._id.toString())
      )

      const favoritesWithProfile = await Promise.all(
        validFavorites.map(async (fav) => {
          const project = fav.project
          let profile = null
          if (project && project.account) {
            if (project.account.role === 'student') {
              const student = await Student.findOne({ account: project.account._id }).select('name')
              profile = student ? { name: student.name } : null
            } else if (project.account.role === 'employer') {
              const employer = await Employer.findOne({ account: project.account._id }).select('companyName')
              profile = employer ? { companyName: employer.companyName } : null
            }
          }
          return {
            ...fav.toObject(),
            project: {
              ...project.toObject(),
              profile
            }
          }
        })
      )

      res.status(200).json(favoritesWithProfile)
    } catch (err) {
      res.status(500).json({ message: 'Error retrieving favorites', error: err.message })
    }
  }

  // [DELETE] /favorites/:id
  async removeFavorite(req, res, next) {
    try {
      const accountId = req.account._id
      const { id } = req.params

      const deletedFavorite = await Favorite.findOneAndDelete({ account: accountId, project: id })
      if (!deletedFavorite) {
        return res.status(404).json({ message: 'Favorite not found' })
      }

      res.status(200).json({ message: 'Favorite removed successfully' })
    } catch (err) {
      res.status(500).json({ message: 'Error removing favorite', error: err.message })
    }
  }
}

module.exports = new FavoriteController()

File: backend.NET/Controllers/FavoriteController.cs
using System.Text.Json;
using api.Models;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;

namespace api.Controllers
{
    [Route("[controller]s")]
    [Authorize]
    [ApiController]
    public class FavoriteController : ControllerBase
    {
        private readonly ILogger<FavoriteController> _logger;
        private readonly JobiverseContext _context;
        public record FavoriteDto(string projectId);

        public FavoriteController(ILogger<FavoriteController> logger, JobiverseContext context)
        {
            _logger = logger;
            _context = context;
        }

        [HttpPost]
        public async Task<IActionResult> SaveFavorite([FromBody] FavoriteDto dto)
        {
            var accountId = User.FindFirst("AccountId")?.Value;
            if (accountId == null)
                return Unauthorized("AccountId not found in token");

            var project = await _context.Projects.FindAsync(dto.projectId);
            if (project == null)
                return NotFound(new { message = "Project not found" });

            var existingFavorite = await _context.Set<Favorite>()
                .FirstOrDefaultAsync(f => f.AccountId == accountId && f.ProjectId == dto.projectId);

            if (existingFavorite != null)
                return BadRequest(new { message = "This project is already in your favorites" });

            var favorite = new Favorite
            {
                AccountId = accountId,
                ProjectId = dto.projectId
            };

            _context.Add(favorite);
            await _context.SaveChangesAsync();

            return StatusCode(201, new { message = "Project saved to favorites successfully", favorite });
        }

        [HttpGet]
        public async Task<IActionResult> GetFavorites()
        {
            var accountId = User.FindFirst("AccountId")?.Value;
            if (accountId == null)
                return Unauthorized("AccountId not found in token");

            var existingProjectIds = await _context.Projects
                .Select(p => p.ProjectId)
                .ToListAsync();

            var favorites = await _context.Set<Favorite>()
                .Where(f => f.AccountId == accountId)
                .Include(f => f.Project)
                    .ThenInclude(p => p.Account)
                .Include(f => f.Project)
                    .ThenInclude(p => p.ProjectLocation)
                .OrderByDescending(f => f.Project.CreatedAt)
                .ToListAsync();

            var validFavorites = favorites
                .Where(f => f.Project != null && existingProjectIds.Contains(f.ProjectId))
                .ToList();

            var result = new List<object>();

            foreach (var fav in validFavorites)
            {
                object? profile = null;
                var project = fav.Project;
                if (project?.Account != null)
                {
                    if (project.Account.AccountRole == "student")
                    {
                        var student = await _context.Set<Student>()
                            .AsNoTracking()
                            .FirstOrDefaultAsync(s => s.AccountId == project.Account.AccountId);
                        if (student != null)
                            profile = new { name = student.Name };
                    }
                    else if (project.Account.AccountRole == "employer")
                    {
                        var employer = await _context.Set<Employer>()
                            .AsNoTracking()
                            .FirstOrDefaultAsync(e => e.AccountId == project.Account.AccountId);
                        if (employer != null)
                            profile = new { companyName = employer.CompanyName };
                    }
                }

                result.Add(new
                {
                    //favoriteId = new { fav.AccountId, fav.ProjectId },
                    fav.CreatedAt,
                    project = new
                    {
                        _id = project?.ProjectId,
                        project?.Title,
                        location = project?.ProjectLocation == null ? null : new
                        {
                            project.ProjectLocation.Province,
                            project.ProjectLocation.District,
                            project.ProjectLocation.Ward
                        },
                        project?.Status,
                        project?.Salary,
                        project?.WorkType,
                        account = project?.Account == null ? null : new
                        {
                            role = project.Account.AccountRole,
                            avatar = project.Account.Avatar == null ? null : new
                            {
                                data = Convert.ToBase64String(project.Account.Avatar),
                                contentType = project.Account.AvatarType
                            }
                        },
                        profile,
                    }
                });
            }

            return Ok(result);
        }

        [HttpDelete("{id}")]
        public async Task<IActionResult> RemoveFavorite(string id)
        {
            var accountId = User.FindFirst("AccountId")?.Value;
            if (accountId == null)
                return Unauthorized("AccountId not found in token");

            var favorite = await _context.Set<Favorite>()
                .FirstOrDefaultAsync(f => f.AccountId == accountId && f.ProjectId == id);

            if (favorite == null)
                return NotFound(new { message = "Favorite not found" });

            _context.Remove(favorite);
            await _context.SaveChangesAsync();

            return Ok(new { message = "Favorite removed successfully" });
        }
    }
}
